### PR TITLE
Update RetrievalTool Gateway to allow user setting retriever/reranker params

### PR DESCRIPTION
**Pull request by minmin-intel**
## Description

Previously user can only pass in query to the retrieval tool and there is no way to set retriever/reranker params with requests to the gateway. This PR is to expose these params to users and allow users to set retriever/reranker params with the requests to the gateway.

## Issues

`n/a`.

## Type of change

List the type of change like below. Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds new functionality)

## Dependencies

No newly introduced 3rd party dependency.

## Tests

DocIndexRetriever example test

### CODE REVIEW

File: comps/cores/mega/gateway.py [hunk 1]
```diff
@@ -622,31 +622,41 @@ def __init__(self, megaservice, host="0.0.0.0", port=8889):
...
-        for key, TypeClass in zip(["text", "input", "input"], [TextDoc, EmbeddingRequest, ChatCompletionRequest]):
-            query = parser_input(data, TypeClass, key)
-            if query is not None:
-                break
-        if query is None:
-            raise ValueError(f"Unknown request type: {data}")
-        result_dict, runtime_graph = await self.megaservice.schedule(initial_inputs={"text": query})
+        chat_request = ChatCompletionRequest.parse_obj(data)
+        query = chat_request.input
+        print("Query: ", query)
```
Comment (xuechendi) [inline comps/cores/mega/gateway.py:633]: need remove print here

File: comps/cores/mega/gateway.py [hunk 2]
```diff
@@ -622,31 +622,41 @@ def __init__(self, megaservice, host="0.0.0.0", port=8889):
             host,
             port,
             str(MegaServiceEndpoint.RETRIEVALTOOL),
-            Union[TextDoc, EmbeddingRequest, ChatCompletionRequest],  # ChatCompletionRequest,
-            Union[RerankedDoc, LLMParamsDoc],  # ChatCompletionResponse
+            ChatCompletionRequest,
```
Comment (xuechendi) [inline comps/cores/mega/gateway.py:625]: Are we not expecting TextDoc / EmbeddingRequest in future retrieval request? @ftian1 @lvliang-intel

File: comps/cores/mega/gateway.py [hunk 3]
```diff
@@ -622,31 +622,41 @@ def __init__(self, megaservice, host="0.0.0.0", port=8889):
...
+            distance_threshold=chat_request.distance_threshold if chat_request.distance_threshold else None,
+            fetch_k=chat_request.fetch_k if chat_request.fetch_k else 20,
+            lambda_mult=chat_request.lambda_mult if chat_request.lambda_mult else 0.5,
+            score_threshold=chat_request.score_threshold if chat_request.score_threshold else 0.2,
+        )
+        reranker_parameters = RerankerParms(
+            top_n=chat_request.top_n if chat_request.top_n else 1,
+        )
+
+        print("Retriever parameters:\n", retriever_parameters)
```
Comment (xuechendi) [inline comps/cores/mega/gateway.py:647]: remove print

File: comps/cores/mega/gateway.py [hunk 4]
```diff
@@ -622,31 +622,41 @@ def __init__(self, megaservice, host="0.0.0.0", port=8889):
...
+        result_dict, runtime_graph = await self.megaservice.schedule(
+            initial_inputs={"messages": "", "input": query, "k": chat_request.k, "top_n": chat_request.top_n},
+            retriever_parameters=retriever_parameters,
+            reranker_parameters=reranker_parameters,
+        )
+
         last_node = runtime_graph.all_leaves()[-1]
         response = result_dict[last_node]
-        print("response is ", response)
+        print("Response:\n", response)
```
Comment (xuechendi) [inline comps/cores/mega/gateway.py:659]: remove print

File: comps/cores/mega/gateway.py (hunk 2 above)
Comment (minmin-intel) [reply comps/cores/mega/gateway.py:625]: Good catch. I think we should also support TextDoc and EmbeddingRequest inputs. Let me add them back in.

File: comps/cores/mega/gateway.py (hunk 4 above)
Comment (minmin-intel) [reply comps/cores/mega/gateway.py]: Done

File: comps/cores/mega/gateway.py (hunk 3 above)
Comment (minmin-intel) [reply comps/cores/mega/gateway.py]: Done.

File: comps/cores/mega/gateway.py (hunk 2 above)
Comment (minmin-intel) [reply comps/cores/mega/gateway.py]: @xuechendi  I added the other input types back. Please review.

**Review by xuechendi — APPROVED** [review]
Codes Looks good to me
My only concern is that it looks like this PR is not covered by any UT?

**Comment by xuechendi** [comment]
@chensuyue , is there any UT will cover gateway code change in GenAIComp?

**Comment by minmin-intel** [comment]
> Codes Looks good to me My only concern is that it looks like this PR is not covered by any UT?

Looks like they pick ChatQnA example as UT. But I agree that the code changes in RetrievalToolGate was not tested in CICD. @chensuyue Can we add DocIndexRetriever example in the Example Test?

**Review by lvliang-intel — APPROVED** [review]

**Pull request merged**